src/lib/storage/constraints/row_templated_constraint_checker.hpp

```diff
@@ -0,0 +1,146 @@
+#include "storage/constraints/base_constraint_checker.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+template <typename Row>
+class RowTemplatedConstraintChecker : public BaseConstraintChecker {
+ public:
+  RowTemplatedConstraintChecker(const Table& table, const TableConstraintDefinition& constraint)
+      : BaseConstraintChecker(table, constraint) {}
+
+  /**
+   * Extracts the rows that are being inserted from the table given to the insert operator.
+   */
+  virtual std::shared_ptr<std::vector<Row>> get_inserted_rows(std::shared_ptr<const Table> table) const = 0;
+
+  /**
+   * Prepare for returning values when get_row is called.
+   * Return false if the segment doesn't need to be checked for duplicate values.
+   */
+  virtual bool preprocess_chunk(std::shared_ptr<const Chunk> chunk) { return true; }
+
+  /**
+   * Returns a row from the current chunk (pre_process_chunk is called before) at the given chunk offset.
+   * Should return a null optional if the row is null or contains a null, i.e. no constraint check is required.
+   */
+  virtual std::optional<Row> get_row(std::shared_ptr<const Chunk> chunk, const ChunkOffset chunk_offset) const = 0;
+
+  /**
+   * Checks if a constraint is still satisfied when some values are inserted into the table.
+   * Snapshot commit ID and transaction ID are required, also a start chunk ID can be passed to skip
+   * compressed chunks on incremental unique checkings during the commit phase.
+   */
+  virtual std::tuple<bool, ChunkID> is_valid_for_inserted_values(std::shared_ptr<const Table> table_to_insert,
+                                                                 const CommitID snapshot_commit_id,
+                                                                 const TransactionID our_tid,
+                                                                 const ChunkID start_chunk_id) {
+    // We store all rows to be inserted in an ordered vector and check for each row in the table
+    // with a binary search if that row is contained in the inserted values.
+    // In almost all cases (different number of inserted values) this is faster than using a set or unordered_set:
+    // http://quick-bench.com/005z63fU43ivxWQsl9Fm_P8zu0k
+    _values_to_insert = get_inserted_rows(table_to_insert);
+    std::sort(_values_to_insert->begin(), _values_to_insert->end());
+
+    // Also, if there is only a single row to be inserted we keep it stored directly for faster access.
+    std::optional<Row> single_insert_value{};
+    if (_values_to_insert->size() == 1) {
+      single_insert_value = (*_values_to_insert)[0];
+    }
+
+    // We remember and return the id of the first mutable chunk.
+    // Reason is that we can skip checking compressed chunks during the commit as they have been checked
+    // already with the operator and won't change later.
+    std::optional<ChunkID> first_mutable_chunk{};
+
+    ChunkID chunk_id{0};
+    for (const auto& chunk : this->_table.chunks()) {
```

Comment:
Here again, if you increment the chunk_id in every case, why not just use a for-loop?